{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "67e8c2fe",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import scipy.io\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "f4c4792f",
   "metadata": {},
   "outputs": [],
   "source": [
    "file_loc = \"../Hao2021/Data/GSE164378_RAW/\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4dea4079",
   "metadata": {},
   "source": [
    "# RNA "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "fcec4e24",
   "metadata": {},
   "outputs": [],
   "source": [
    "RNA_matrix = scipy.io.mmread(file_loc+\"GSM5008737_RNA_3P-matrix.mtx.gz\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "cde6a6e5",
   "metadata": {},
   "outputs": [],
   "source": [
    "RNA_matrix = RNA_matrix.tocsr()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "e9b80101",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<161764x33538 sparse matrix of type '<class 'numpy.int64'>'\n",
       "\twith 354321830 stored elements in Compressed Sparse Column format>"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "RNA_matrix.T"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "id": "789dfb23",
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>L1_AAACCCAAGAAACTCA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>L1_AAACCCAAGACATACA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>L1_AAACCCACAACTGGTT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>L1_AAACCCACACGTACTA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>L1_AAACCCACAGCATACT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>161759</th>\n",
       "      <td>E2L8_TTTGTTGGTCGTGATT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>161760</th>\n",
       "      <td>E2L8_TTTGTTGGTGTGCCTG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>161761</th>\n",
       "      <td>E2L8_TTTGTTGGTTAGTTCG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>161762</th>\n",
       "      <td>E2L8_TTTGTTGGTTGGCTAT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>161763</th>\n",
       "      <td>E2L8_TTTGTTGTCTCATGGA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>161764 rows × 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                            0\n",
       "0         L1_AAACCCAAGAAACTCA\n",
       "1         L1_AAACCCAAGACATACA\n",
       "2         L1_AAACCCACAACTGGTT\n",
       "3         L1_AAACCCACACGTACTA\n",
       "4         L1_AAACCCACAGCATACT\n",
       "...                       ...\n",
       "161759  E2L8_TTTGTTGGTCGTGATT\n",
       "161760  E2L8_TTTGTTGGTGTGCCTG\n",
       "161761  E2L8_TTTGTTGGTTAGTTCG\n",
       "161762  E2L8_TTTGTTGGTTGGCTAT\n",
       "161763  E2L8_TTTGTTGTCTCATGGA\n",
       "\n",
       "[161764 rows x 1 columns]"
      ]
     },
     "execution_count": 59,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "RNA_barcodes_3P = pd.read_csv(file_loc+\"GSM5008739_HTO_3P-barcodes.tsv.gz\", \n",
    "                              sep = \"\\t\", header = None)\n",
    "RNA_barcodes_3P"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "3587fb8e",
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(33538, 3)"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "RNA_features_3P = pd.read_csv(\"../Hao2021/Data/GSE164378_RAW/GSM5008737_RNA_3P-features.tsv.gz\", \n",
    "                              sep = \"\\t\", header = None)\n",
    "RNA_features_3P.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f868a251",
   "metadata": {},
   "source": [
    "# ADT\n",
    "A panel of antibodies"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "ca873467",
   "metadata": {},
   "outputs": [],
   "source": [
    "ADT_counts = scipy.io.mmread(file_loc+\"GSM5008738_ADT_3P-matrix.mtx.gz\")\n",
    "ADT_counts = ADT_counts.tocsr()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "fb5ea688",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<228x161764 sparse matrix of type '<class 'numpy.int64'>'\n",
       "\twith 33599680 stored elements in COOrdinate format>"
      ]
     },
     "execution_count": 54,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ADT_counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "fbc1ac7c",
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Barcode</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>L1_AAACCCAAGAAACTCA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>L1_AAACCCAAGACATACA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>L1_AAACCCACAACTGGTT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>L1_AAACCCACACGTACTA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>L1_AAACCCACAGCATACT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>161759</th>\n",
       "      <td>E2L8_TTTGTTGGTCGTGATT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>161760</th>\n",
       "      <td>E2L8_TTTGTTGGTGTGCCTG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>161761</th>\n",
       "      <td>E2L8_TTTGTTGGTTAGTTCG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>161762</th>\n",
       "      <td>E2L8_TTTGTTGGTTGGCTAT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>161763</th>\n",
       "      <td>E2L8_TTTGTTGTCTCATGGA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>161764 rows × 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                      Barcode\n",
       "0         L1_AAACCCAAGAAACTCA\n",
       "1         L1_AAACCCAAGACATACA\n",
       "2         L1_AAACCCACAACTGGTT\n",
       "3         L1_AAACCCACACGTACTA\n",
       "4         L1_AAACCCACAGCATACT\n",
       "...                       ...\n",
       "161759  E2L8_TTTGTTGGTCGTGATT\n",
       "161760  E2L8_TTTGTTGGTGTGCCTG\n",
       "161761  E2L8_TTTGTTGGTTAGTTCG\n",
       "161762  E2L8_TTTGTTGGTTGGCTAT\n",
       "161763  E2L8_TTTGTTGTCTCATGGA\n",
       "\n",
       "[161764 rows x 1 columns]"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ADT_barcodes = pd.read_csv(file_loc+\"GSM5008738_ADT_3P-barcodes.tsv.gz\",\n",
    "                          sep = '\\t', header = None, names = ['Barcode'])\n",
    "# ADT_barcodes.shape\n",
    "ADT_barcodes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "c0a85826",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Finding the unique identifiers from the barcodes: [identifer]_[barcode]\n",
    "ADT_barcodes[['Identifier', 'Barcode' ]] = ADT_barcodes['Barcode'].str.split('_', expand = True )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "745939a5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['L1', 'L2', 'L3', 'L4', 'L5', 'E2L1', 'E2L2', 'E2L3', 'E2L4',\n",
       "       'E2L5', 'E2L6', 'E2L7', 'E2L8'], dtype=object)"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ADT_barcodes['Identifier'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "7cef49b0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(228, 3)"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ADT_features_3P = pd.read_csv(file_loc+\"GSM5008738_ADT_3P-features.tsv.gz\", \n",
    "                              sep = '\\t', header = None)\n",
    "ADT_features_3P.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "4a42ee68",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "228"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ADT"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "id": "8e5ac8df",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 64,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ADT_features_3P[0].equals(ADT_features_3P[1]) "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0f2b9a9c",
   "metadata": {},
   "source": [
    "# HTO\n",
    "Hash-tag oligo"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "51e22a63",
   "metadata": {},
   "outputs": [],
   "source": [
    "HTO_matrix = scipy.io.mmread(file_loc+\"GSM5008739_HTO_3P-matrix.mtx.gz\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "id": "cc0c2b3d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<24x161764 sparse matrix of type '<class 'numpy.int64'>'\n",
       "\twith 1893738 stored elements in COOrdinate format>"
      ]
     },
     "execution_count": 75,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "HTO_matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "id": "96dd20c2",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(161764, 1)"
      ]
     },
     "execution_count": 74,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "HTO_barcodes_3P = pd.read_csv(file_loc+\"GSM5008739_HTO_3P-barcodes.tsv.gz\",\n",
    "                         sep='\\t', header = None)\n",
    "HTO_barcodes_3P.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "78bb330b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(24, 3)"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "HTO_features_3P = pd.read_csv(file_loc+\"GSM5008739_HTO_3P-features.tsv.gz\", \n",
    "                              sep = '\\t', header = None)\n",
    "HTO_features_3P.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "9434d969",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Batch1-HTO10</td>\n",
       "      <td>Batch1-HTO10</td>\n",
       "      <td>Gene Expression</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Batch1-HTO13</td>\n",
       "      <td>Batch1-HTO13</td>\n",
       "      <td>Gene Expression</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Batch1-HTO12</td>\n",
       "      <td>Batch1-HTO12</td>\n",
       "      <td>Gene Expression</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Batch1-HTO5</td>\n",
       "      <td>Batch1-HTO5</td>\n",
       "      <td>Gene Expression</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Batch1-HTO4</td>\n",
       "      <td>Batch1-HTO4</td>\n",
       "      <td>Gene Expression</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Batch1-HTO7</td>\n",
       "      <td>Batch1-HTO7</td>\n",
       "      <td>Gene Expression</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Batch1-HTO6</td>\n",
       "      <td>Batch1-HTO6</td>\n",
       "      <td>Gene Expression</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Batch1-HTO1</td>\n",
       "      <td>Batch1-HTO1</td>\n",
       "      <td>Gene Expression</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Batch1-HTO3</td>\n",
       "      <td>Batch1-HTO3</td>\n",
       "      <td>Gene Expression</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Batch1-HTO2</td>\n",
       "      <td>Batch1-HTO2</td>\n",
       "      <td>Gene Expression</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>Batch1-HTO9</td>\n",
       "      <td>Batch1-HTO9</td>\n",
       "      <td>Gene Expression</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>Batch1-HTO8</td>\n",
       "      <td>Batch1-HTO8</td>\n",
       "      <td>Gene Expression</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>Batch2-HTO10</td>\n",
       "      <td>Batch2-HTO10</td>\n",
       "      <td>Gene Expression</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>Batch2-HTO13</td>\n",
       "      <td>Batch2-HTO13</td>\n",
       "      <td>Gene Expression</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>Batch2-HTO12</td>\n",
       "      <td>Batch2-HTO12</td>\n",
       "      <td>Gene Expression</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>Batch2-HTO5</td>\n",
       "      <td>Batch2-HTO5</td>\n",
       "      <td>Gene Expression</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>Batch2-HTO4</td>\n",
       "      <td>Batch2-HTO4</td>\n",
       "      <td>Gene Expression</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>Batch2-HTO7</td>\n",
       "      <td>Batch2-HTO7</td>\n",
       "      <td>Gene Expression</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>Batch2-HTO6</td>\n",
       "      <td>Batch2-HTO6</td>\n",
       "      <td>Gene Expression</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>Batch2-HTO1</td>\n",
       "      <td>Batch2-HTO1</td>\n",
       "      <td>Gene Expression</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>Batch2-HTO3</td>\n",
       "      <td>Batch2-HTO3</td>\n",
       "      <td>Gene Expression</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>Batch2-HTO2</td>\n",
       "      <td>Batch2-HTO2</td>\n",
       "      <td>Gene Expression</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>Batch2-HTO9</td>\n",
       "      <td>Batch2-HTO9</td>\n",
       "      <td>Gene Expression</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>Batch2-HTO8</td>\n",
       "      <td>Batch2-HTO8</td>\n",
       "      <td>Gene Expression</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               0             1                2\n",
       "0   Batch1-HTO10  Batch1-HTO10  Gene Expression\n",
       "1   Batch1-HTO13  Batch1-HTO13  Gene Expression\n",
       "2   Batch1-HTO12  Batch1-HTO12  Gene Expression\n",
       "3    Batch1-HTO5   Batch1-HTO5  Gene Expression\n",
       "4    Batch1-HTO4   Batch1-HTO4  Gene Expression\n",
       "5    Batch1-HTO7   Batch1-HTO7  Gene Expression\n",
       "6    Batch1-HTO6   Batch1-HTO6  Gene Expression\n",
       "7    Batch1-HTO1   Batch1-HTO1  Gene Expression\n",
       "8    Batch1-HTO3   Batch1-HTO3  Gene Expression\n",
       "9    Batch1-HTO2   Batch1-HTO2  Gene Expression\n",
       "10   Batch1-HTO9   Batch1-HTO9  Gene Expression\n",
       "11   Batch1-HTO8   Batch1-HTO8  Gene Expression\n",
       "12  Batch2-HTO10  Batch2-HTO10  Gene Expression\n",
       "13  Batch2-HTO13  Batch2-HTO13  Gene Expression\n",
       "14  Batch2-HTO12  Batch2-HTO12  Gene Expression\n",
       "15   Batch2-HTO5   Batch2-HTO5  Gene Expression\n",
       "16   Batch2-HTO4   Batch2-HTO4  Gene Expression\n",
       "17   Batch2-HTO7   Batch2-HTO7  Gene Expression\n",
       "18   Batch2-HTO6   Batch2-HTO6  Gene Expression\n",
       "19   Batch2-HTO1   Batch2-HTO1  Gene Expression\n",
       "20   Batch2-HTO3   Batch2-HTO3  Gene Expression\n",
       "21   Batch2-HTO2   Batch2-HTO2  Gene Expression\n",
       "22   Batch2-HTO9   Batch2-HTO9  Gene Expression\n",
       "23   Batch2-HTO8   Batch2-HTO8  Gene Expression"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "HTO_features_3P"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "id": "40e65019",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 68,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "HTO_features_3P[0].equals(HTO_features_3P[1])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4c6a0d13",
   "metadata": {},
   "source": [
    "# Metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7e22466e",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc_meta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "c3243abf",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(2393, 18)"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "TCRab_batch1_annotations = pd.read_csv(\"../Hao2021/Data/GSE164378_RAW/GSM5008743_TCRab_batch1_annotations.csv.gz\")\n",
    "TCR_gb_batch1_annotations.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "caa16e6f",
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>barcode</th>\n",
       "      <th>is_cell</th>\n",
       "      <th>contig_id</th>\n",
       "      <th>high_confidence</th>\n",
       "      <th>length</th>\n",
       "      <th>chain</th>\n",
       "      <th>v_gene</th>\n",
       "      <th>d_gene</th>\n",
       "      <th>j_gene</th>\n",
       "      <th>c_gene</th>\n",
       "      <th>full_length</th>\n",
       "      <th>productive</th>\n",
       "      <th>cdr3</th>\n",
       "      <th>cdr3_nt</th>\n",
       "      <th>reads</th>\n",
       "      <th>umis</th>\n",
       "      <th>raw_clonotype_id</th>\n",
       "      <th>raw_consensus_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AAACCTGAGAACTCGG-1</td>\n",
       "      <td>True</td>\n",
       "      <td>AAACCTGAGAACTCGG-1_contig_1</td>\n",
       "      <td>True</td>\n",
       "      <td>526</td>\n",
       "      <td>TRB</td>\n",
       "      <td>TRBV11-3</td>\n",
       "      <td>TRBD2</td>\n",
       "      <td>TRBJ2-1</td>\n",
       "      <td>TRBC2</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>CASSLEPGASSYNEQFF</td>\n",
       "      <td>TGTGCCAGCAGCTTAGAACCGGGAGCCAGCTCCTACAATGAGCAGT...</td>\n",
       "      <td>1891</td>\n",
       "      <td>5</td>\n",
       "      <td>clonotype153</td>\n",
       "      <td>clonotype153_consensus_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AAACCTGAGAACTCGG-1</td>\n",
       "      <td>True</td>\n",
       "      <td>AAACCTGAGAACTCGG-1_contig_3</td>\n",
       "      <td>True</td>\n",
       "      <td>503</td>\n",
       "      <td>TRA</td>\n",
       "      <td>TRAV9-2</td>\n",
       "      <td>None</td>\n",
       "      <td>TRAJ52</td>\n",
       "      <td>TRAC</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>CALFNAGGTSYGKLTF</td>\n",
       "      <td>TGTGCTCTTTTTAATGCTGGTGGTACTAGCTATGGAAAGCTGACATTT</td>\n",
       "      <td>592</td>\n",
       "      <td>1</td>\n",
       "      <td>clonotype153</td>\n",
       "      <td>clonotype153_consensus_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AAACCTGAGTGGAGAA-1</td>\n",
       "      <td>True</td>\n",
       "      <td>AAACCTGAGTGGAGAA-1_contig_1</td>\n",
       "      <td>True</td>\n",
       "      <td>506</td>\n",
       "      <td>TRB</td>\n",
       "      <td>TRBV5-1</td>\n",
       "      <td>None</td>\n",
       "      <td>TRBJ2-3</td>\n",
       "      <td>TRBC2</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>CASSSRGTDTQYF</td>\n",
       "      <td>TGCGCCAGCAGCTCCCGGGGCACAGATACGCAGTATTTT</td>\n",
       "      <td>1297</td>\n",
       "      <td>5</td>\n",
       "      <td>clonotype154</td>\n",
       "      <td>clonotype154_consensus_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AAACCTGAGTGGAGAA-1</td>\n",
       "      <td>True</td>\n",
       "      <td>AAACCTGAGTGGAGAA-1_contig_2</td>\n",
       "      <td>True</td>\n",
       "      <td>653</td>\n",
       "      <td>TRB</td>\n",
       "      <td>TRBV23-1</td>\n",
       "      <td>None</td>\n",
       "      <td>TRBJ2-6</td>\n",
       "      <td>TRBC2</td>\n",
       "      <td>True</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>729</td>\n",
       "      <td>7</td>\n",
       "      <td>clonotype154</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AAACCTGAGTGGAGAA-1</td>\n",
       "      <td>True</td>\n",
       "      <td>AAACCTGAGTGGAGAA-1_contig_3</td>\n",
       "      <td>True</td>\n",
       "      <td>522</td>\n",
       "      <td>TRA</td>\n",
       "      <td>TRAV13-1</td>\n",
       "      <td>None</td>\n",
       "      <td>TRAJ21</td>\n",
       "      <td>TRAC</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>CAARDFNKFYF</td>\n",
       "      <td>TGTGCAGCGCGGGACTTCAACAAATTTTACTTT</td>\n",
       "      <td>577</td>\n",
       "      <td>2</td>\n",
       "      <td>clonotype154</td>\n",
       "      <td>clonotype154_consensus_2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              barcode  is_cell                    contig_id  high_confidence  \\\n",
       "0  AAACCTGAGAACTCGG-1     True  AAACCTGAGAACTCGG-1_contig_1             True   \n",
       "1  AAACCTGAGAACTCGG-1     True  AAACCTGAGAACTCGG-1_contig_3             True   \n",
       "2  AAACCTGAGTGGAGAA-1     True  AAACCTGAGTGGAGAA-1_contig_1             True   \n",
       "3  AAACCTGAGTGGAGAA-1     True  AAACCTGAGTGGAGAA-1_contig_2             True   \n",
       "4  AAACCTGAGTGGAGAA-1     True  AAACCTGAGTGGAGAA-1_contig_3             True   \n",
       "\n",
       "   length chain    v_gene d_gene   j_gene c_gene  full_length productive  \\\n",
       "0     526   TRB  TRBV11-3  TRBD2  TRBJ2-1  TRBC2         True       True   \n",
       "1     503   TRA   TRAV9-2   None   TRAJ52   TRAC         True       True   \n",
       "2     506   TRB   TRBV5-1   None  TRBJ2-3  TRBC2         True       True   \n",
       "3     653   TRB  TRBV23-1   None  TRBJ2-6  TRBC2         True       None   \n",
       "4     522   TRA  TRAV13-1   None   TRAJ21   TRAC         True       True   \n",
       "\n",
       "                cdr3                                            cdr3_nt  \\\n",
       "0  CASSLEPGASSYNEQFF  TGTGCCAGCAGCTTAGAACCGGGAGCCAGCTCCTACAATGAGCAGT...   \n",
       "1   CALFNAGGTSYGKLTF   TGTGCTCTTTTTAATGCTGGTGGTACTAGCTATGGAAAGCTGACATTT   \n",
       "2      CASSSRGTDTQYF            TGCGCCAGCAGCTCCCGGGGCACAGATACGCAGTATTTT   \n",
       "3               None                                               None   \n",
       "4        CAARDFNKFYF                  TGTGCAGCGCGGGACTTCAACAAATTTTACTTT   \n",
       "\n",
       "   reads  umis raw_clonotype_id          raw_consensus_id  \n",
       "0   1891     5     clonotype153  clonotype153_consensus_2  \n",
       "1    592     1     clonotype153  clonotype153_consensus_1  \n",
       "2   1297     5     clonotype154  clonotype154_consensus_1  \n",
       "3    729     7     clonotype154                      None  \n",
       "4    577     2     clonotype154  clonotype154_consensus_2  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "TCRab_batch1_annotations.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "af1b1689",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>barcode</th>\n",
       "      <th>is_cell</th>\n",
       "      <th>contig_id</th>\n",
       "      <th>high_confidence</th>\n",
       "      <th>length</th>\n",
       "      <th>chain</th>\n",
       "      <th>v_gene</th>\n",
       "      <th>d_gene</th>\n",
       "      <th>j_gene</th>\n",
       "      <th>c_gene</th>\n",
       "      <th>full_length</th>\n",
       "      <th>productive</th>\n",
       "      <th>cdr3</th>\n",
       "      <th>cdr3_nt</th>\n",
       "      <th>reads</th>\n",
       "      <th>umis</th>\n",
       "      <th>raw_clonotype_id</th>\n",
       "      <th>raw_consensus_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AAACCTGCAAATCCGT-1</td>\n",
       "      <td>True</td>\n",
       "      <td>AAACCTGCAAATCCGT-1_contig_1</td>\n",
       "      <td>True</td>\n",
       "      <td>661</td>\n",
       "      <td>TRG</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>TRGJP1</td>\n",
       "      <td>TRGC1</td>\n",
       "      <td>False</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>4185</td>\n",
       "      <td>2</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AAACCTGTCGGAGGTA-1</td>\n",
       "      <td>True</td>\n",
       "      <td>AAACCTGTCGGAGGTA-1_contig_1</td>\n",
       "      <td>True</td>\n",
       "      <td>764</td>\n",
       "      <td>TRG</td>\n",
       "      <td>TRGV2</td>\n",
       "      <td>None</td>\n",
       "      <td>TRGJ2</td>\n",
       "      <td>TRGC2</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>CATWDGPNYYKKLF</td>\n",
       "      <td>TGTGCCACCTGGGACGGGCCGAATTATTATAAGAAACTCTTT</td>\n",
       "      <td>10339</td>\n",
       "      <td>3</td>\n",
       "      <td>clonotype20</td>\n",
       "      <td>clonotype20_consensus_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AAACGGGTCCGCATAA-1</td>\n",
       "      <td>True</td>\n",
       "      <td>AAACGGGTCCGCATAA-1_contig_1</td>\n",
       "      <td>True</td>\n",
       "      <td>559</td>\n",
       "      <td>Multi</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>TRDJ1</td>\n",
       "      <td>TRDC</td>\n",
       "      <td>False</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>14745</td>\n",
       "      <td>4</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AAACGGGTCTATCGCC-1</td>\n",
       "      <td>True</td>\n",
       "      <td>AAACGGGTCTATCGCC-1_contig_1</td>\n",
       "      <td>True</td>\n",
       "      <td>509</td>\n",
       "      <td>Multi</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>TRDJ1</td>\n",
       "      <td>TRDC</td>\n",
       "      <td>False</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>11350</td>\n",
       "      <td>2</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AAACGGGTCTATCGCC-1</td>\n",
       "      <td>True</td>\n",
       "      <td>AAACGGGTCTATCGCC-1_contig_2</td>\n",
       "      <td>True</td>\n",
       "      <td>341</td>\n",
       "      <td>TRG</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>TRGJP1</td>\n",
       "      <td>TRGC1</td>\n",
       "      <td>False</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>5576</td>\n",
       "      <td>2</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              barcode  is_cell                    contig_id  high_confidence  \\\n",
       "0  AAACCTGCAAATCCGT-1     True  AAACCTGCAAATCCGT-1_contig_1             True   \n",
       "1  AAACCTGTCGGAGGTA-1     True  AAACCTGTCGGAGGTA-1_contig_1             True   \n",
       "2  AAACGGGTCCGCATAA-1     True  AAACGGGTCCGCATAA-1_contig_1             True   \n",
       "3  AAACGGGTCTATCGCC-1     True  AAACGGGTCTATCGCC-1_contig_1             True   \n",
       "4  AAACGGGTCTATCGCC-1     True  AAACGGGTCTATCGCC-1_contig_2             True   \n",
       "\n",
       "   length  chain v_gene d_gene  j_gene c_gene  full_length productive  \\\n",
       "0     661    TRG   None   None  TRGJP1  TRGC1        False       None   \n",
       "1     764    TRG  TRGV2   None   TRGJ2  TRGC2         True       True   \n",
       "2     559  Multi   None   None   TRDJ1   TRDC        False       None   \n",
       "3     509  Multi   None   None   TRDJ1   TRDC        False       None   \n",
       "4     341    TRG   None   None  TRGJP1  TRGC1        False       None   \n",
       "\n",
       "             cdr3                                     cdr3_nt  reads  umis  \\\n",
       "0            None                                        None   4185     2   \n",
       "1  CATWDGPNYYKKLF  TGTGCCACCTGGGACGGGCCGAATTATTATAAGAAACTCTTT  10339     3   \n",
       "2            None                                        None  14745     4   \n",
       "3            None                                        None  11350     2   \n",
       "4            None                                        None   5576     2   \n",
       "\n",
       "  raw_clonotype_id         raw_consensus_id  \n",
       "0             None                     None  \n",
       "1      clonotype20  clonotype20_consensus_1  \n",
       "2             None                     None  \n",
       "3             None                     None  \n",
       "4             None                     None  "
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "TCR_gb_batch1_annotations = pd.read_csv(\"../Hao2021/Data/GSE164378_RAW/GSM5008744_TCRgd_batch1_annotations.csv.gz\")\n",
    "TCR_gb_batch1_annotations.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "5f389bcc",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(2393, 18)"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "TCR_gb_batch1_annotations.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "0654e5a5",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc_metadata_3p = pd.read_csv(\"../Hao2021/Data/GSE164378_sc.meta.data_3P.csv.gz\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "b063f5e7",
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>nCount_ADT</th>\n",
       "      <th>nFeature_ADT</th>\n",
       "      <th>nCount_RNA</th>\n",
       "      <th>nFeature_RNA</th>\n",
       "      <th>orig.ident</th>\n",
       "      <th>lane</th>\n",
       "      <th>donor</th>\n",
       "      <th>time</th>\n",
       "      <th>celltype.l1</th>\n",
       "      <th>celltype.l2</th>\n",
       "      <th>celltype.l3</th>\n",
       "      <th>Phase</th>\n",
       "      <th>Batch</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>L1_AAACCCAAGAAACTCA</td>\n",
       "      <td>7535</td>\n",
       "      <td>217</td>\n",
       "      <td>10823</td>\n",
       "      <td>2915</td>\n",
       "      <td>SeuratProject</td>\n",
       "      <td>L1</td>\n",
       "      <td>P2</td>\n",
       "      <td>7</td>\n",
       "      <td>Mono</td>\n",
       "      <td>CD14 Mono</td>\n",
       "      <td>CD14 Mono</td>\n",
       "      <td>G1</td>\n",
       "      <td>Batch1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>L1_AAACCCAAGACATACA</td>\n",
       "      <td>6013</td>\n",
       "      <td>209</td>\n",
       "      <td>5864</td>\n",
       "      <td>1617</td>\n",
       "      <td>SeuratProject</td>\n",
       "      <td>L1</td>\n",
       "      <td>P1</td>\n",
       "      <td>7</td>\n",
       "      <td>CD4 T</td>\n",
       "      <td>CD4 TCM</td>\n",
       "      <td>CD4 TCM_1</td>\n",
       "      <td>G1</td>\n",
       "      <td>Batch1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>L1_AAACCCACAACTGGTT</td>\n",
       "      <td>6620</td>\n",
       "      <td>213</td>\n",
       "      <td>5067</td>\n",
       "      <td>1381</td>\n",
       "      <td>SeuratProject</td>\n",
       "      <td>L1</td>\n",
       "      <td>P4</td>\n",
       "      <td>2</td>\n",
       "      <td>CD8 T</td>\n",
       "      <td>CD8 Naive</td>\n",
       "      <td>CD8 Naive</td>\n",
       "      <td>S</td>\n",
       "      <td>Batch1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>L1_AAACCCACACGTACTA</td>\n",
       "      <td>3567</td>\n",
       "      <td>202</td>\n",
       "      <td>4786</td>\n",
       "      <td>1890</td>\n",
       "      <td>SeuratProject</td>\n",
       "      <td>L1</td>\n",
       "      <td>P3</td>\n",
       "      <td>7</td>\n",
       "      <td>NK</td>\n",
       "      <td>NK</td>\n",
       "      <td>NK_2</td>\n",
       "      <td>G1</td>\n",
       "      <td>Batch1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>L1_AAACCCACAGCATACT</td>\n",
       "      <td>6402</td>\n",
       "      <td>215</td>\n",
       "      <td>6505</td>\n",
       "      <td>1621</td>\n",
       "      <td>SeuratProject</td>\n",
       "      <td>L1</td>\n",
       "      <td>P4</td>\n",
       "      <td>7</td>\n",
       "      <td>CD8 T</td>\n",
       "      <td>CD8 Naive</td>\n",
       "      <td>CD8 Naive</td>\n",
       "      <td>G1</td>\n",
       "      <td>Batch1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>L1_AAACCCACATCAGTCA</td>\n",
       "      <td>5297</td>\n",
       "      <td>212</td>\n",
       "      <td>4332</td>\n",
       "      <td>1633</td>\n",
       "      <td>SeuratProject</td>\n",
       "      <td>L1</td>\n",
       "      <td>P3</td>\n",
       "      <td>2</td>\n",
       "      <td>CD8 T</td>\n",
       "      <td>CD8 TEM</td>\n",
       "      <td>CD8 TEM_1</td>\n",
       "      <td>G1</td>\n",
       "      <td>Batch1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>L1_AAACCCACATCGGTTA</td>\n",
       "      <td>7634</td>\n",
       "      <td>219</td>\n",
       "      <td>9333</td>\n",
       "      <td>2672</td>\n",
       "      <td>SeuratProject</td>\n",
       "      <td>L1</td>\n",
       "      <td>P1</td>\n",
       "      <td>7</td>\n",
       "      <td>Mono</td>\n",
       "      <td>CD16 Mono</td>\n",
       "      <td>CD16 Mono</td>\n",
       "      <td>G1</td>\n",
       "      <td>Batch1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>L1_AAACCCACATGGATCT</td>\n",
       "      <td>8210</td>\n",
       "      <td>222</td>\n",
       "      <td>3589</td>\n",
       "      <td>1122</td>\n",
       "      <td>SeuratProject</td>\n",
       "      <td>L1</td>\n",
       "      <td>P4</td>\n",
       "      <td>2</td>\n",
       "      <td>B</td>\n",
       "      <td>B intermediate</td>\n",
       "      <td>B intermediate lambda</td>\n",
       "      <td>G1</td>\n",
       "      <td>Batch1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>L1_AAACCCAGTGGAACAC</td>\n",
       "      <td>2847</td>\n",
       "      <td>201</td>\n",
       "      <td>3251</td>\n",
       "      <td>1375</td>\n",
       "      <td>SeuratProject</td>\n",
       "      <td>L1</td>\n",
       "      <td>P1</td>\n",
       "      <td>2</td>\n",
       "      <td>NK</td>\n",
       "      <td>NK</td>\n",
       "      <td>NK_2</td>\n",
       "      <td>G2M</td>\n",
       "      <td>Batch1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>L1_AAACCCATCCACACCT</td>\n",
       "      <td>4557</td>\n",
       "      <td>209</td>\n",
       "      <td>3401</td>\n",
       "      <td>1200</td>\n",
       "      <td>SeuratProject</td>\n",
       "      <td>L1</td>\n",
       "      <td>P3</td>\n",
       "      <td>2</td>\n",
       "      <td>CD8 T</td>\n",
       "      <td>CD8 Naive</td>\n",
       "      <td>CD8 Naive</td>\n",
       "      <td>S</td>\n",
       "      <td>Batch1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>L1_AAACCCATCTCCCTAG</td>\n",
       "      <td>6557</td>\n",
       "      <td>215</td>\n",
       "      <td>6244</td>\n",
       "      <td>1568</td>\n",
       "      <td>SeuratProject</td>\n",
       "      <td>L1</td>\n",
       "      <td>P3</td>\n",
       "      <td>0</td>\n",
       "      <td>CD4 T</td>\n",
       "      <td>CD4 Naive</td>\n",
       "      <td>CD4 Naive</td>\n",
       "      <td>G1</td>\n",
       "      <td>Batch1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>L1_AAACCCATCTGCGGAC</td>\n",
       "      <td>5129</td>\n",
       "      <td>212</td>\n",
       "      <td>4198</td>\n",
       "      <td>1318</td>\n",
       "      <td>SeuratProject</td>\n",
       "      <td>L1</td>\n",
       "      <td>P1</td>\n",
       "      <td>0</td>\n",
       "      <td>CD4 T</td>\n",
       "      <td>CD4 TCM</td>\n",
       "      <td>CD4 TCM_1</td>\n",
       "      <td>S</td>\n",
       "      <td>Batch1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>L1_AAACCCATCTTAAGGC</td>\n",
       "      <td>5382</td>\n",
       "      <td>215</td>\n",
       "      <td>4016</td>\n",
       "      <td>1451</td>\n",
       "      <td>SeuratProject</td>\n",
       "      <td>L1</td>\n",
       "      <td>P2</td>\n",
       "      <td>2</td>\n",
       "      <td>CD4 T</td>\n",
       "      <td>CD4 CTL</td>\n",
       "      <td>CD4 CTL</td>\n",
       "      <td>S</td>\n",
       "      <td>Batch1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>L1_AAACGAAAGATAACAC</td>\n",
       "      <td>5155</td>\n",
       "      <td>216</td>\n",
       "      <td>3393</td>\n",
       "      <td>1092</td>\n",
       "      <td>SeuratProject</td>\n",
       "      <td>L1</td>\n",
       "      <td>P2</td>\n",
       "      <td>2</td>\n",
       "      <td>B</td>\n",
       "      <td>B naive</td>\n",
       "      <td>B naive kappa</td>\n",
       "      <td>S</td>\n",
       "      <td>Batch1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>L1_AAACGAAAGGCCCAAA</td>\n",
       "      <td>8989</td>\n",
       "      <td>221</td>\n",
       "      <td>6866</td>\n",
       "      <td>2000</td>\n",
       "      <td>SeuratProject</td>\n",
       "      <td>L1</td>\n",
       "      <td>P4</td>\n",
       "      <td>0</td>\n",
       "      <td>Mono</td>\n",
       "      <td>CD14 Mono</td>\n",
       "      <td>CD14 Mono</td>\n",
       "      <td>G1</td>\n",
       "      <td>Batch1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>L1_AAACGAAAGTTACTCG</td>\n",
       "      <td>7630</td>\n",
       "      <td>208</td>\n",
       "      <td>5486</td>\n",
       "      <td>1390</td>\n",
       "      <td>SeuratProject</td>\n",
       "      <td>L1</td>\n",
       "      <td>P1</td>\n",
       "      <td>0</td>\n",
       "      <td>CD4 T</td>\n",
       "      <td>CD4 TCM</td>\n",
       "      <td>CD4 TCM_3</td>\n",
       "      <td>G1</td>\n",
       "      <td>Batch1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>L1_AAACGAAAGTTGAATG</td>\n",
       "      <td>3978</td>\n",
       "      <td>210</td>\n",
       "      <td>3543</td>\n",
       "      <td>1255</td>\n",
       "      <td>SeuratProject</td>\n",
       "      <td>L1</td>\n",
       "      <td>P3</td>\n",
       "      <td>7</td>\n",
       "      <td>other T</td>\n",
       "      <td>MAIT</td>\n",
       "      <td>MAIT</td>\n",
       "      <td>S</td>\n",
       "      <td>Batch1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>L1_AAACGAACAAGAGTAT</td>\n",
       "      <td>3344</td>\n",
       "      <td>208</td>\n",
       "      <td>3989</td>\n",
       "      <td>1582</td>\n",
       "      <td>SeuratProject</td>\n",
       "      <td>L1</td>\n",
       "      <td>P2</td>\n",
       "      <td>7</td>\n",
       "      <td>NK</td>\n",
       "      <td>NK</td>\n",
       "      <td>NK_2</td>\n",
       "      <td>G2M</td>\n",
       "      <td>Batch1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>L1_AAACGAACAATGAGCG</td>\n",
       "      <td>5465</td>\n",
       "      <td>211</td>\n",
       "      <td>7379</td>\n",
       "      <td>2315</td>\n",
       "      <td>SeuratProject</td>\n",
       "      <td>L1</td>\n",
       "      <td>P1</td>\n",
       "      <td>7</td>\n",
       "      <td>CD4 T</td>\n",
       "      <td>CD4 TCM</td>\n",
       "      <td>CD4 TCM_2</td>\n",
       "      <td>G1</td>\n",
       "      <td>Batch1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>L1_AAACGAACACAGCATT</td>\n",
       "      <td>5008</td>\n",
       "      <td>218</td>\n",
       "      <td>4387</td>\n",
       "      <td>1724</td>\n",
       "      <td>SeuratProject</td>\n",
       "      <td>L1</td>\n",
       "      <td>P3</td>\n",
       "      <td>0</td>\n",
       "      <td>NK</td>\n",
       "      <td>NK</td>\n",
       "      <td>NK_2</td>\n",
       "      <td>S</td>\n",
       "      <td>Batch1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>L1_AAACGAACACTCCTTG</td>\n",
       "      <td>3114</td>\n",
       "      <td>206</td>\n",
       "      <td>4444</td>\n",
       "      <td>1878</td>\n",
       "      <td>SeuratProject</td>\n",
       "      <td>L1</td>\n",
       "      <td>P1</td>\n",
       "      <td>2</td>\n",
       "      <td>NK</td>\n",
       "      <td>NK</td>\n",
       "      <td>NK_2</td>\n",
       "      <td>S</td>\n",
       "      <td>Batch1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>L1_AAACGAACAGCATCTA</td>\n",
       "      <td>5922</td>\n",
       "      <td>211</td>\n",
       "      <td>5693</td>\n",
       "      <td>1629</td>\n",
       "      <td>SeuratProject</td>\n",
       "      <td>L1</td>\n",
       "      <td>P1</td>\n",
       "      <td>2</td>\n",
       "      <td>CD4 T</td>\n",
       "      <td>CD4 TCM</td>\n",
       "      <td>CD4 TCM_1</td>\n",
       "      <td>S</td>\n",
       "      <td>Batch1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>L1_AAACGAACAGCTACTA</td>\n",
       "      <td>3161</td>\n",
       "      <td>211</td>\n",
       "      <td>1538</td>\n",
       "      <td>822</td>\n",
       "      <td>SeuratProject</td>\n",
       "      <td>L1</td>\n",
       "      <td>P4</td>\n",
       "      <td>7</td>\n",
       "      <td>NK</td>\n",
       "      <td>NK</td>\n",
       "      <td>NK_2</td>\n",
       "      <td>G1</td>\n",
       "      <td>Batch1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>L1_AAACGAAGTATTCTCT</td>\n",
       "      <td>5236</td>\n",
       "      <td>210</td>\n",
       "      <td>4747</td>\n",
       "      <td>1397</td>\n",
       "      <td>SeuratProject</td>\n",
       "      <td>L1</td>\n",
       "      <td>P3</td>\n",
       "      <td>0</td>\n",
       "      <td>CD4 T</td>\n",
       "      <td>CD4 TCM</td>\n",
       "      <td>CD4 TCM_3</td>\n",
       "      <td>S</td>\n",
       "      <td>Batch1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>L1_AAACGAAGTCACAGAG</td>\n",
       "      <td>5044</td>\n",
       "      <td>208</td>\n",
       "      <td>5354</td>\n",
       "      <td>1519</td>\n",
       "      <td>SeuratProject</td>\n",
       "      <td>L1</td>\n",
       "      <td>P3</td>\n",
       "      <td>7</td>\n",
       "      <td>CD4 T</td>\n",
       "      <td>CD4 TCM</td>\n",
       "      <td>CD4 TCM_1</td>\n",
       "      <td>G2M</td>\n",
       "      <td>Batch1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>L1_AAACGAATCCCGTTGT</td>\n",
       "      <td>6940</td>\n",
       "      <td>207</td>\n",
       "      <td>4200</td>\n",
       "      <td>1392</td>\n",
       "      <td>SeuratProject</td>\n",
       "      <td>L1</td>\n",
       "      <td>P2</td>\n",
       "      <td>2</td>\n",
       "      <td>CD8 T</td>\n",
       "      <td>CD8 TEM</td>\n",
       "      <td>CD8 TEM_2</td>\n",
       "      <td>G1</td>\n",
       "      <td>Batch1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>L1_AAACGAATCCTCACCA</td>\n",
       "      <td>3192</td>\n",
       "      <td>203</td>\n",
       "      <td>4525</td>\n",
       "      <td>1459</td>\n",
       "      <td>SeuratProject</td>\n",
       "      <td>L1</td>\n",
       "      <td>P1</td>\n",
       "      <td>2</td>\n",
       "      <td>other T</td>\n",
       "      <td>gdT</td>\n",
       "      <td>gdT_3</td>\n",
       "      <td>G1</td>\n",
       "      <td>Batch1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>L1_AAACGAATCGTTCTCG</td>\n",
       "      <td>4978</td>\n",
       "      <td>211</td>\n",
       "      <td>3943</td>\n",
       "      <td>1178</td>\n",
       "      <td>SeuratProject</td>\n",
       "      <td>L1</td>\n",
       "      <td>P2</td>\n",
       "      <td>2</td>\n",
       "      <td>CD4 T</td>\n",
       "      <td>CD4 Naive</td>\n",
       "      <td>CD4 Naive</td>\n",
       "      <td>S</td>\n",
       "      <td>Batch1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>L1_AAACGAATCTCACTCG</td>\n",
       "      <td>3005</td>\n",
       "      <td>204</td>\n",
       "      <td>3519</td>\n",
       "      <td>1544</td>\n",
       "      <td>SeuratProject</td>\n",
       "      <td>L1</td>\n",
       "      <td>P1</td>\n",
       "      <td>2</td>\n",
       "      <td>NK</td>\n",
       "      <td>NK</td>\n",
       "      <td>NK_1</td>\n",
       "      <td>G1</td>\n",
       "      <td>Batch1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>L1_AAACGCTAGAGCATTA</td>\n",
       "      <td>6164</td>\n",
       "      <td>205</td>\n",
       "      <td>4382</td>\n",
       "      <td>1426</td>\n",
       "      <td>SeuratProject</td>\n",
       "      <td>L1</td>\n",
       "      <td>P2</td>\n",
       "      <td>7</td>\n",
       "      <td>CD8 T</td>\n",
       "      <td>CD8 TEM</td>\n",
       "      <td>CD8 TEM_2</td>\n",
       "      <td>G2M</td>\n",
       "      <td>Batch1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>L1_AAACGCTAGGAAAGGT</td>\n",
       "      <td>6871</td>\n",
       "      <td>209</td>\n",
       "      <td>5869</td>\n",
       "      <td>1425</td>\n",
       "      <td>SeuratProject</td>\n",
       "      <td>L1</td>\n",
       "      <td>P2</td>\n",
       "      <td>2</td>\n",
       "      <td>CD8 T</td>\n",
       "      <td>CD8 Naive</td>\n",
       "      <td>CD8 Naive</td>\n",
       "      <td>G1</td>\n",
       "      <td>Batch1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>L1_AAACGCTAGGATATGT</td>\n",
       "      <td>5726</td>\n",
       "      <td>206</td>\n",
       "      <td>6473</td>\n",
       "      <td>1517</td>\n",
       "      <td>SeuratProject</td>\n",
       "      <td>L1</td>\n",
       "      <td>P2</td>\n",
       "      <td>0</td>\n",
       "      <td>CD4 T</td>\n",
       "      <td>CD4 Naive</td>\n",
       "      <td>CD4 Naive</td>\n",
       "      <td>G1</td>\n",
       "      <td>Batch1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>L1_AAACGCTAGGTCGTCC</td>\n",
       "      <td>7332</td>\n",
       "      <td>208</td>\n",
       "      <td>5120</td>\n",
       "      <td>1688</td>\n",
       "      <td>SeuratProject</td>\n",
       "      <td>L1</td>\n",
       "      <td>P1</td>\n",
       "      <td>0</td>\n",
       "      <td>CD8 T</td>\n",
       "      <td>CD8 TEM</td>\n",
       "      <td>CD8 TEM_2</td>\n",
       "      <td>S</td>\n",
       "      <td>Batch1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>L1_AAACGCTAGTACCCTA</td>\n",
       "      <td>7022</td>\n",
       "      <td>213</td>\n",
       "      <td>7311</td>\n",
       "      <td>2330</td>\n",
       "      <td>SeuratProject</td>\n",
       "      <td>L1</td>\n",
       "      <td>P3</td>\n",
       "      <td>7</td>\n",
       "      <td>Mono</td>\n",
       "      <td>CD14 Mono</td>\n",
       "      <td>CD14 Mono</td>\n",
       "      <td>G1</td>\n",
       "      <td>Batch1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>L1_AAACGCTCAACGATCT</td>\n",
       "      <td>4665</td>\n",
       "      <td>205</td>\n",
       "      <td>4593</td>\n",
       "      <td>1435</td>\n",
       "      <td>SeuratProject</td>\n",
       "      <td>L1</td>\n",
       "      <td>P3</td>\n",
       "      <td>0</td>\n",
       "      <td>CD8 T</td>\n",
       "      <td>CD8 TCM</td>\n",
       "      <td>CD8 TCM_1</td>\n",
       "      <td>S</td>\n",
       "      <td>Batch1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>L1_AAACGCTCACTAACGT</td>\n",
       "      <td>5222</td>\n",
       "      <td>213</td>\n",
       "      <td>6621</td>\n",
       "      <td>1482</td>\n",
       "      <td>SeuratProject</td>\n",
       "      <td>L1</td>\n",
       "      <td>P1</td>\n",
       "      <td>2</td>\n",
       "      <td>CD4 T</td>\n",
       "      <td>CD4 Naive</td>\n",
       "      <td>CD4 Naive</td>\n",
       "      <td>S</td>\n",
       "      <td>Batch1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>L1_AAACGCTCACTAGAGG</td>\n",
       "      <td>8305</td>\n",
       "      <td>218</td>\n",
       "      <td>12616</td>\n",
       "      <td>3274</td>\n",
       "      <td>SeuratProject</td>\n",
       "      <td>L1</td>\n",
       "      <td>P4</td>\n",
       "      <td>2</td>\n",
       "      <td>Mono</td>\n",
       "      <td>CD14 Mono</td>\n",
       "      <td>CD14 Mono</td>\n",
       "      <td>G1</td>\n",
       "      <td>Batch1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>L1_AAACGCTCATAGGTTC</td>\n",
       "      <td>7868</td>\n",
       "      <td>215</td>\n",
       "      <td>9353</td>\n",
       "      <td>2679</td>\n",
       "      <td>SeuratProject</td>\n",
       "      <td>L1</td>\n",
       "      <td>P3</td>\n",
       "      <td>0</td>\n",
       "      <td>Mono</td>\n",
       "      <td>CD16 Mono</td>\n",
       "      <td>CD16 Mono</td>\n",
       "      <td>S</td>\n",
       "      <td>Batch1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>L1_AAACGCTGTCCGGATC</td>\n",
       "      <td>5375</td>\n",
       "      <td>202</td>\n",
       "      <td>5118</td>\n",
       "      <td>1384</td>\n",
       "      <td>SeuratProject</td>\n",
       "      <td>L1</td>\n",
       "      <td>P3</td>\n",
       "      <td>7</td>\n",
       "      <td>CD4 T</td>\n",
       "      <td>CD4 Naive</td>\n",
       "      <td>CD4 Naive</td>\n",
       "      <td>S</td>\n",
       "      <td>Batch1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>L1_AAACGCTGTGCTCGTG</td>\n",
       "      <td>2272</td>\n",
       "      <td>203</td>\n",
       "      <td>2369</td>\n",
       "      <td>1062</td>\n",
       "      <td>SeuratProject</td>\n",
       "      <td>L1</td>\n",
       "      <td>P1</td>\n",
       "      <td>2</td>\n",
       "      <td>other T</td>\n",
       "      <td>dnT</td>\n",
       "      <td>dnT_2</td>\n",
       "      <td>G1</td>\n",
       "      <td>Batch1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>L1_AAACGCTTCCACAAGT</td>\n",
       "      <td>6137</td>\n",
       "      <td>206</td>\n",
       "      <td>5059</td>\n",
       "      <td>1482</td>\n",
       "      <td>SeuratProject</td>\n",
       "      <td>L1</td>\n",
       "      <td>P2</td>\n",
       "      <td>0</td>\n",
       "      <td>CD8 T</td>\n",
       "      <td>CD8 Naive</td>\n",
       "      <td>CD8 Naive</td>\n",
       "      <td>S</td>\n",
       "      <td>Batch1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>L1_AAACGCTTCTTGGTCC</td>\n",
       "      <td>5159</td>\n",
       "      <td>215</td>\n",
       "      <td>4135</td>\n",
       "      <td>1321</td>\n",
       "      <td>SeuratProject</td>\n",
       "      <td>L1</td>\n",
       "      <td>P1</td>\n",
       "      <td>7</td>\n",
       "      <td>B</td>\n",
       "      <td>B intermediate</td>\n",
       "      <td>B intermediate lambda</td>\n",
       "      <td>G1</td>\n",
       "      <td>Batch1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>L1_AAAGAACAGAGGTGCT</td>\n",
       "      <td>2876</td>\n",
       "      <td>208</td>\n",
       "      <td>3296</td>\n",
       "      <td>1413</td>\n",
       "      <td>SeuratProject</td>\n",
       "      <td>L1</td>\n",
       "      <td>P3</td>\n",
       "      <td>0</td>\n",
       "      <td>NK</td>\n",
       "      <td>NK</td>\n",
       "      <td>NK_2</td>\n",
       "      <td>S</td>\n",
       "      <td>Batch1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>L1_AAAGAACAGCGACTGA</td>\n",
       "      <td>5140</td>\n",
       "      <td>207</td>\n",
       "      <td>5169</td>\n",
       "      <td>1269</td>\n",
       "      <td>SeuratProject</td>\n",
       "      <td>L1</td>\n",
       "      <td>P2</td>\n",
       "      <td>7</td>\n",
       "      <td>CD4 T</td>\n",
       "      <td>CD4 Naive</td>\n",
       "      <td>CD4 Naive</td>\n",
       "      <td>S</td>\n",
       "      <td>Batch1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>L1_AAAGAACAGGTTCACT</td>\n",
       "      <td>6056</td>\n",
       "      <td>202</td>\n",
       "      <td>4967</td>\n",
       "      <td>1332</td>\n",
       "      <td>SeuratProject</td>\n",
       "      <td>L1</td>\n",
       "      <td>P4</td>\n",
       "      <td>7</td>\n",
       "      <td>CD4 T</td>\n",
       "      <td>CD4 TCM</td>\n",
       "      <td>CD4 TCM_1</td>\n",
       "      <td>G2M</td>\n",
       "      <td>Batch1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>L1_AAAGAACCAAGCGGAT</td>\n",
       "      <td>5998</td>\n",
       "      <td>205</td>\n",
       "      <td>5084</td>\n",
       "      <td>1359</td>\n",
       "      <td>SeuratProject</td>\n",
       "      <td>L1</td>\n",
       "      <td>P4</td>\n",
       "      <td>7</td>\n",
       "      <td>CD4 T</td>\n",
       "      <td>CD4 TCM</td>\n",
       "      <td>CD4 TCM_3</td>\n",
       "      <td>S</td>\n",
       "      <td>Batch1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>L1_AAAGAACCAAGGAGTC</td>\n",
       "      <td>5398</td>\n",
       "      <td>211</td>\n",
       "      <td>4389</td>\n",
       "      <td>1206</td>\n",
       "      <td>SeuratProject</td>\n",
       "      <td>L1</td>\n",
       "      <td>P2</td>\n",
       "      <td>0</td>\n",
       "      <td>B</td>\n",
       "      <td>B intermediate</td>\n",
       "      <td>B intermediate kappa</td>\n",
       "      <td>S</td>\n",
       "      <td>Batch1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>L1_AAAGAACCACCCTTGT</td>\n",
       "      <td>5443</td>\n",
       "      <td>212</td>\n",
       "      <td>2864</td>\n",
       "      <td>900</td>\n",
       "      <td>SeuratProject</td>\n",
       "      <td>L1</td>\n",
       "      <td>P2</td>\n",
       "      <td>2</td>\n",
       "      <td>CD4 T</td>\n",
       "      <td>CD4 Naive</td>\n",
       "      <td>CD4 Naive</td>\n",
       "      <td>S</td>\n",
       "      <td>Batch1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>L1_AAAGAACCACCTCTAC</td>\n",
       "      <td>6178</td>\n",
       "      <td>216</td>\n",
       "      <td>5079</td>\n",
       "      <td>1494</td>\n",
       "      <td>SeuratProject</td>\n",
       "      <td>L1</td>\n",
       "      <td>P4</td>\n",
       "      <td>0</td>\n",
       "      <td>B</td>\n",
       "      <td>B memory</td>\n",
       "      <td>B memory kappa</td>\n",
       "      <td>G1</td>\n",
       "      <td>Batch1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49</th>\n",
       "      <td>L1_AAAGAACCACGCGCAT</td>\n",
       "      <td>10565</td>\n",
       "      <td>220</td>\n",
       "      <td>8127</td>\n",
       "      <td>2368</td>\n",
       "      <td>SeuratProject</td>\n",
       "      <td>L1</td>\n",
       "      <td>P1</td>\n",
       "      <td>7</td>\n",
       "      <td>Mono</td>\n",
       "      <td>CD14 Mono</td>\n",
       "      <td>CD14 Mono</td>\n",
       "      <td>G1</td>\n",
       "      <td>Batch1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             Unnamed: 0  nCount_ADT  nFeature_ADT  nCount_RNA  nFeature_RNA  \\\n",
       "0   L1_AAACCCAAGAAACTCA        7535           217       10823          2915   \n",
       "1   L1_AAACCCAAGACATACA        6013           209        5864          1617   \n",
       "2   L1_AAACCCACAACTGGTT        6620           213        5067          1381   \n",
       "3   L1_AAACCCACACGTACTA        3567           202        4786          1890   \n",
       "4   L1_AAACCCACAGCATACT        6402           215        6505          1621   \n",
       "5   L1_AAACCCACATCAGTCA        5297           212        4332          1633   \n",
       "6   L1_AAACCCACATCGGTTA        7634           219        9333          2672   \n",
       "7   L1_AAACCCACATGGATCT        8210           222        3589          1122   \n",
       "8   L1_AAACCCAGTGGAACAC        2847           201        3251          1375   \n",
       "9   L1_AAACCCATCCACACCT        4557           209        3401          1200   \n",
       "10  L1_AAACCCATCTCCCTAG        6557           215        6244          1568   \n",
       "11  L1_AAACCCATCTGCGGAC        5129           212        4198          1318   \n",
       "12  L1_AAACCCATCTTAAGGC        5382           215        4016          1451   \n",
       "13  L1_AAACGAAAGATAACAC        5155           216        3393          1092   \n",
       "14  L1_AAACGAAAGGCCCAAA        8989           221        6866          2000   \n",
       "15  L1_AAACGAAAGTTACTCG        7630           208        5486          1390   \n",
       "16  L1_AAACGAAAGTTGAATG        3978           210        3543          1255   \n",
       "17  L1_AAACGAACAAGAGTAT        3344           208        3989          1582   \n",
       "18  L1_AAACGAACAATGAGCG        5465           211        7379          2315   \n",
       "19  L1_AAACGAACACAGCATT        5008           218        4387          1724   \n",
       "20  L1_AAACGAACACTCCTTG        3114           206        4444          1878   \n",
       "21  L1_AAACGAACAGCATCTA        5922           211        5693          1629   \n",
       "22  L1_AAACGAACAGCTACTA        3161           211        1538           822   \n",
       "23  L1_AAACGAAGTATTCTCT        5236           210        4747          1397   \n",
       "24  L1_AAACGAAGTCACAGAG        5044           208        5354          1519   \n",
       "25  L1_AAACGAATCCCGTTGT        6940           207        4200          1392   \n",
       "26  L1_AAACGAATCCTCACCA        3192           203        4525          1459   \n",
       "27  L1_AAACGAATCGTTCTCG        4978           211        3943          1178   \n",
       "28  L1_AAACGAATCTCACTCG        3005           204        3519          1544   \n",
       "29  L1_AAACGCTAGAGCATTA        6164           205        4382          1426   \n",
       "30  L1_AAACGCTAGGAAAGGT        6871           209        5869          1425   \n",
       "31  L1_AAACGCTAGGATATGT        5726           206        6473          1517   \n",
       "32  L1_AAACGCTAGGTCGTCC        7332           208        5120          1688   \n",
       "33  L1_AAACGCTAGTACCCTA        7022           213        7311          2330   \n",
       "34  L1_AAACGCTCAACGATCT        4665           205        4593          1435   \n",
       "35  L1_AAACGCTCACTAACGT        5222           213        6621          1482   \n",
       "36  L1_AAACGCTCACTAGAGG        8305           218       12616          3274   \n",
       "37  L1_AAACGCTCATAGGTTC        7868           215        9353          2679   \n",
       "38  L1_AAACGCTGTCCGGATC        5375           202        5118          1384   \n",
       "39  L1_AAACGCTGTGCTCGTG        2272           203        2369          1062   \n",
       "40  L1_AAACGCTTCCACAAGT        6137           206        5059          1482   \n",
       "41  L1_AAACGCTTCTTGGTCC        5159           215        4135          1321   \n",
       "42  L1_AAAGAACAGAGGTGCT        2876           208        3296          1413   \n",
       "43  L1_AAAGAACAGCGACTGA        5140           207        5169          1269   \n",
       "44  L1_AAAGAACAGGTTCACT        6056           202        4967          1332   \n",
       "45  L1_AAAGAACCAAGCGGAT        5998           205        5084          1359   \n",
       "46  L1_AAAGAACCAAGGAGTC        5398           211        4389          1206   \n",
       "47  L1_AAAGAACCACCCTTGT        5443           212        2864           900   \n",
       "48  L1_AAAGAACCACCTCTAC        6178           216        5079          1494   \n",
       "49  L1_AAAGAACCACGCGCAT       10565           220        8127          2368   \n",
       "\n",
       "       orig.ident lane donor  time celltype.l1     celltype.l2  \\\n",
       "0   SeuratProject   L1    P2     7        Mono       CD14 Mono   \n",
       "1   SeuratProject   L1    P1     7       CD4 T         CD4 TCM   \n",
       "2   SeuratProject   L1    P4     2       CD8 T       CD8 Naive   \n",
       "3   SeuratProject   L1    P3     7          NK              NK   \n",
       "4   SeuratProject   L1    P4     7       CD8 T       CD8 Naive   \n",
       "5   SeuratProject   L1    P3     2       CD8 T         CD8 TEM   \n",
       "6   SeuratProject   L1    P1     7        Mono       CD16 Mono   \n",
       "7   SeuratProject   L1    P4     2           B  B intermediate   \n",
       "8   SeuratProject   L1    P1     2          NK              NK   \n",
       "9   SeuratProject   L1    P3     2       CD8 T       CD8 Naive   \n",
       "10  SeuratProject   L1    P3     0       CD4 T       CD4 Naive   \n",
       "11  SeuratProject   L1    P1     0       CD4 T         CD4 TCM   \n",
       "12  SeuratProject   L1    P2     2       CD4 T         CD4 CTL   \n",
       "13  SeuratProject   L1    P2     2           B         B naive   \n",
       "14  SeuratProject   L1    P4     0        Mono       CD14 Mono   \n",
       "15  SeuratProject   L1    P1     0       CD4 T         CD4 TCM   \n",
       "16  SeuratProject   L1    P3     7     other T            MAIT   \n",
       "17  SeuratProject   L1    P2     7          NK              NK   \n",
       "18  SeuratProject   L1    P1     7       CD4 T         CD4 TCM   \n",
       "19  SeuratProject   L1    P3     0          NK              NK   \n",
       "20  SeuratProject   L1    P1     2          NK              NK   \n",
       "21  SeuratProject   L1    P1     2       CD4 T         CD4 TCM   \n",
       "22  SeuratProject   L1    P4     7          NK              NK   \n",
       "23  SeuratProject   L1    P3     0       CD4 T         CD4 TCM   \n",
       "24  SeuratProject   L1    P3     7       CD4 T         CD4 TCM   \n",
       "25  SeuratProject   L1    P2     2       CD8 T         CD8 TEM   \n",
       "26  SeuratProject   L1    P1     2     other T             gdT   \n",
       "27  SeuratProject   L1    P2     2       CD4 T       CD4 Naive   \n",
       "28  SeuratProject   L1    P1     2          NK              NK   \n",
       "29  SeuratProject   L1    P2     7       CD8 T         CD8 TEM   \n",
       "30  SeuratProject   L1    P2     2       CD8 T       CD8 Naive   \n",
       "31  SeuratProject   L1    P2     0       CD4 T       CD4 Naive   \n",
       "32  SeuratProject   L1    P1     0       CD8 T         CD8 TEM   \n",
       "33  SeuratProject   L1    P3     7        Mono       CD14 Mono   \n",
       "34  SeuratProject   L1    P3     0       CD8 T         CD8 TCM   \n",
       "35  SeuratProject   L1    P1     2       CD4 T       CD4 Naive   \n",
       "36  SeuratProject   L1    P4     2        Mono       CD14 Mono   \n",
       "37  SeuratProject   L1    P3     0        Mono       CD16 Mono   \n",
       "38  SeuratProject   L1    P3     7       CD4 T       CD4 Naive   \n",
       "39  SeuratProject   L1    P1     2     other T             dnT   \n",
       "40  SeuratProject   L1    P2     0       CD8 T       CD8 Naive   \n",
       "41  SeuratProject   L1    P1     7           B  B intermediate   \n",
       "42  SeuratProject   L1    P3     0          NK              NK   \n",
       "43  SeuratProject   L1    P2     7       CD4 T       CD4 Naive   \n",
       "44  SeuratProject   L1    P4     7       CD4 T         CD4 TCM   \n",
       "45  SeuratProject   L1    P4     7       CD4 T         CD4 TCM   \n",
       "46  SeuratProject   L1    P2     0           B  B intermediate   \n",
       "47  SeuratProject   L1    P2     2       CD4 T       CD4 Naive   \n",
       "48  SeuratProject   L1    P4     0           B        B memory   \n",
       "49  SeuratProject   L1    P1     7        Mono       CD14 Mono   \n",
       "\n",
       "              celltype.l3 Phase   Batch  \n",
       "0               CD14 Mono    G1  Batch1  \n",
       "1               CD4 TCM_1    G1  Batch1  \n",
       "2               CD8 Naive     S  Batch1  \n",
       "3                    NK_2    G1  Batch1  \n",
       "4               CD8 Naive    G1  Batch1  \n",
       "5               CD8 TEM_1    G1  Batch1  \n",
       "6               CD16 Mono    G1  Batch1  \n",
       "7   B intermediate lambda    G1  Batch1  \n",
       "8                    NK_2   G2M  Batch1  \n",
       "9               CD8 Naive     S  Batch1  \n",
       "10              CD4 Naive    G1  Batch1  \n",
       "11              CD4 TCM_1     S  Batch1  \n",
       "12                CD4 CTL     S  Batch1  \n",
       "13          B naive kappa     S  Batch1  \n",
       "14              CD14 Mono    G1  Batch1  \n",
       "15              CD4 TCM_3    G1  Batch1  \n",
       "16                   MAIT     S  Batch1  \n",
       "17                   NK_2   G2M  Batch1  \n",
       "18              CD4 TCM_2    G1  Batch1  \n",
       "19                   NK_2     S  Batch1  \n",
       "20                   NK_2     S  Batch1  \n",
       "21              CD4 TCM_1     S  Batch1  \n",
       "22                   NK_2    G1  Batch1  \n",
       "23              CD4 TCM_3     S  Batch1  \n",
       "24              CD4 TCM_1   G2M  Batch1  \n",
       "25              CD8 TEM_2    G1  Batch1  \n",
       "26                  gdT_3    G1  Batch1  \n",
       "27              CD4 Naive     S  Batch1  \n",
       "28                   NK_1    G1  Batch1  \n",
       "29              CD8 TEM_2   G2M  Batch1  \n",
       "30              CD8 Naive    G1  Batch1  \n",
       "31              CD4 Naive    G1  Batch1  \n",
       "32              CD8 TEM_2     S  Batch1  \n",
       "33              CD14 Mono    G1  Batch1  \n",
       "34              CD8 TCM_1     S  Batch1  \n",
       "35              CD4 Naive     S  Batch1  \n",
       "36              CD14 Mono    G1  Batch1  \n",
       "37              CD16 Mono     S  Batch1  \n",
       "38              CD4 Naive     S  Batch1  \n",
       "39                  dnT_2    G1  Batch1  \n",
       "40              CD8 Naive     S  Batch1  \n",
       "41  B intermediate lambda    G1  Batch1  \n",
       "42                   NK_2     S  Batch1  \n",
       "43              CD4 Naive     S  Batch1  \n",
       "44              CD4 TCM_1   G2M  Batch1  \n",
       "45              CD4 TCM_3     S  Batch1  \n",
       "46   B intermediate kappa     S  Batch1  \n",
       "47              CD4 Naive     S  Batch1  \n",
       "48         B memory kappa    G1  Batch1  \n",
       "49              CD14 Mono    G1  Batch1  "
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sc_metadata_3p.head(50)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "29bb5200",
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(161764, 14)"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sc_metadata_3p.shape"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
